### PR TITLE
[WIP] Add gRPC support for querying

### PR DESCRIPTION
**Pull request by DirectXMan12**
This sketches out a new "Querying" service to the set of Prometheus v2
gRPC endpoints.  It supports all the read operations of the v1 querying API.

It does not expose itself via the gRPC-web gateway, since the v1 API
handles that itself, and the gateway generator does not entirely
understand how to convert our objects into GET requests.

### CODE REVIEW

**Comment by DirectXMan12** [comment]
@fabxc @brancz per our conversation a while ago, this cuts down my querying time dramatically on large data sets.  I'll get some more concrete numbers for you tomorrow.  Is this something you're willing to accept?  If so, I'll do a bit of cleanup, and get a PR together for a nicer client as well.

**Comment by brancz** [comment]
I'd generally be happy to have gRPC endpoint for querying. For the concrete implementation @brian-brazil and @tomwilkie should have a look as well. Maybe @brian-brazil and @tomwilkie can comment on the general opinion for an additional querying endpoint. Of course we'd want the duplication to be at a minimum.

This PR also needs a rebase.

**Comment by brian-brazil** [comment]
V2 needs to support json also, and it should preferably look exactly like the existing endpoints. The support costs of two mildly different querying endpoints would be quite high, we already get enough confusion from the single version we have.

**Comment by DirectXMan12** [comment]
ack, I'll try to get it working.  I don't think the gateway is going to support automated generation, though...  I can do a bit of digging to see if I can get it working, but there might be some manual endpoint work going on for a couple of the endpoints.

**Comment by DirectXMan12** [comment]
(plus the way that the gateway returns errors is different from the way that the v1 endpoint returns errors, so we'd have to decide if we want all v2 return objects to look the same, or if we want both v1 and v2 querying to look the same)

**Comment by tomwilkie** [comment]
I was hoping to use the v2/gRPC implementation as an opportunity to use gRPC streaming, but I guess that can wait - its not something I'm going to be able to do anytime soon.  I will take a look at this PR tomorrow.

File: prompb/types.proto [hunk 1]
```diff
@@ -49,3 +49,30 @@ message LabelMatcher {
   string name  = 2;
   string value = 3;
 }
+
+message SeriesMatcher {
+  string name = 1;
+  repeated LabelMatcher labels = 2;
```
Comment (tomwilkie) [inline prompb/types.proto]: `LabelMatcher` includes a `name` field already.

File: web/api/v2/admin_api.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+	}
+	return mint, maxt, nil
+}
+
+var (
+	minTime = time.Unix(math.MinInt64/1000+62135596801, 0)
+	maxTime = time.Unix(math.MaxInt64/1000-62135596801, 999999999)
+)
+
+func labelsToProto(lset labels.Labels) pb.Labels {
```
Comment (tomwilkie) [inline web/api/v2/admin_api.go]: `labelsToLabelsProto` exists in storage/remote/codec.go

File: web/api/v2/admin_api.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+	"golang.org/x/net/context"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/pkg/timestamp"
+	pb "github.com/prometheus/prometheus/prompb"
```
Comment (tomwilkie) [inline web/api/v2/admin_api.go]: Is this renaming necessary?

**Comment by tomwilkie** [comment]
> I was hoping to use the v2/gRPC implementation...

Ah I see this is not for remote read but rather for query...

File: prompb/rpc.proto [hunk 4]
```diff
@@ -61,3 +62,56 @@ message SeriesDeleteRequest {
...
+message LabelValuesRequest {
+  string name = 1;
+}
+message LabelValuesResponse {
+  repeated string values = 1;
+}
+
+message QueryAtTimeRequest {
+  google.protobuf.Timestamp time   = 1 [(gogoproto.stdtime) = true];
+  google.protobuf.Duration timeout = 2 [(gogoproto.stdduration) = true];
```
Comment (brian-brazil) [inline prompb/rpc.proto]: Should this still be here with grpc? Seconds are also what the existing api uses.

Comment (tomwilkie) [reply prompb/rpc.proto]: I was wondering about that.  In the remote apis we take ms as an int64; in the delete series admin endpoint we use a Timestamp.

I think we should use a timestamp here, as its more 'semantic', and just ensure the json encoding accepts seconds as a float (or whatever the existing api is).

**Comment by fabxc** [comment]
> V2 needs to support json also, and it should preferably look exactly like the existing endpoints. The support costs of two mildly different querying endpoints would be quite high, we already get enough confusion from the single version we have.

I don't understand this point. There's no need for a new version if it is _exactly_ the same as the previous version.

**Comment by brian-brazil** [comment]
> I don't understand this point. There's no need for a new version if it is exactly the same as the previous version.

There are differences in v2, such as the addition of the new admin apis. From my standpoint a user needs to be able to use only the v2 api to do everything they need, and v1 would be pure legacy. Given that, the transition for existing v1 users needs to be as seamless as possible.

**Comment by Gouthamve** [comment]
Well v2 should have everything and even I think it should look similar to v1, but not a strict requirement.

Having said that, I would oppose using the gateway for the actual JSON API itself as the error handling _sucks_ and actually points you in the other direction. I would prefer if we could have a function per endpoint which is called from both the grpc and JSON handlers.

Essentially the handlers have the logic to serialise and deserialise the data and pass the parameters to the function. This smells an awful lot like go-kit, btw :)

**Comment by krasi-georgiev** [comment]
I also think that `go kit` is worth exploring as I think it will reduce the support cost
@peterbourgon mentioned in his video that different protocol handling is one of the main strengths of go kit - 
https://www.youtube.com/watch?v=JXEjAwNWays&feature=youtu.be&t=1458

the only drawback is that grpc  doesn't support streaming.

I really want to refactor the whole web package using go kit even with the risk of not being siutable for this case I thin it is worth the attempt.

**Comment by brian-brazil** [comment]
> this cuts down my querying time dramatically on large data sets.

Can you give an idea what sort of improvement we're talking about?

**Comment by krasi-georgiev** [comment]
After the dev discussions at Promcon there are no objections that there would be no benefits from streaming in remote read and too complicated to be worth it in the promql so not sure what should we do with this PR.

**Comment by brancz** [comment]
> no benefits from streaming in remote read

That was not the conclusion, just that similar to how the Thanos store API does it would be more beneficial.

**Comment by DirectXMan12** [comment]
This can be closed per the discussion here and some offline discussion.  We can always open a new one if things change.

**Pull request closed without merging**